Library/Client.ts

```diff
@@ -38,269 +45,99 @@ class Client {
-        data.baseType = Contracts.DataTypes.REMOTE_DEPENDENCY;
-        data.baseData = remoteDependency;
-        this.track(data, tagOverrides, contextObjects);
+        this.track(telemetry, Contracts.DataTypes.REMOTE_DEPENDENCY);
     }
 
     /**
      * Immediately send all queued telemetry.
      */
-    public sendPendingData(callback?: (v: string) => void) {
+    public flush(callback?: (v: string) => void) {
         this.channel.triggerSend(false, callback);
     }
 
-    public getEnvelope(
-        data: Contracts.Data<Contracts.Domain>,
-        tagOverrides?: { [key: string]: string; }): Contracts.Envelope {
-        if (Contracts.domainSupportsProperties(data.baseData)) { // Do instanceof check. TS will automatically cast and allow the properties property
-            if (data && data.baseData) {
-                // if no properties are specified just add the common ones
-                if (!data.baseData.properties) {
-                    data.baseData.properties = this.commonProperties;
-                } else {
-                    // otherwise, check each of the common ones
-                    for (var name in this.commonProperties) {
-                        // only override if the property `name` has not been set on this item
-                        if (!data.baseData.properties[name]) {
-                            data.baseData.properties[name] = this.commonProperties[name];
-                        }
-                    }
-                }
-            }
-
-            // sanitize properties
-            data.baseData.properties = Util.validateStringMap(data.baseData.properties);
-        }
-
-        var iKey = this.config.instrumentationKey;
-        var envelope = new Contracts.Envelope();
-        envelope.data = data;
-        envelope.iKey = iKey;
-
-        // this is kind of a hack, but the envelope name is always the same as the data name sans the chars "data"
-        envelope.name =
-            "Microsoft.ApplicationInsights." +
-            iKey.replace(/-/g, "") +
-            "." +
-            data.baseType.substr(0, data.baseType.length - 4);
-        envelope.tags = this.getTags(tagOverrides);
-        envelope.time = (new Date()).toISOString();
-        envelope.ver = 1;
-        envelope.sampleRate = this.config.samplingPercentage;
-
-        return envelope;
-    }
-
     /**
      * Generic track method for all telemetry types
      * @param data the telemetry to send
-     * @param tagOverrides the context tags to use for this telemetry which overwrite default context values
+     * @param telemetryType specify the type of telemetry you are tracking from the list of Contracts.DataTypes
      */
-    public track(
-        data: Contracts.Data<Contracts.Domain>,
-        tagOverrides?: { [key: string]: string; },
-        contextObjects?: { [name: string]: any; }) {
+    public track(telemetry: Telemetry, telemetryType: string) {
```

Comment:
Do we have a better type to restrict here than `string`?

Reply:
Would it help to make `Contracts.DataTypes` an enum?

Reply:
Done